### ISSUE TITLE
Err in char_rnn tutorial

### ISSUE THREAD

**Issue by williamFalcon** [pytorch/tutorials#193]
@apaszke    
This tutorial "[char_rnn_classification](http://pytorch.org/tutorials/intermediate/char_rnn_classification_tutorial.html)" has a bug in the forward part of this code:   

```python    
import torch.nn as nn
from torch.autograd import Variable

class RNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(RNN, self).__init__()

        self.hidden_size = hidden_size

        self.i2h = nn.Linear(input_size + hidden_size, hidden_size)
        self.i2o = nn.Linear(input_size + hidden_size, output_size)
        self.softmax = nn.LogSoftmax(dim=1)

    def forward(self, input, hidden):
        combined = torch.cat((input, hidden), 1)
        hidden = self.i2h(combined)
        output = self.i2o(combined)
        output = self.softmax(output)
        return output, hidden

    def initHidden(self):
        return Variable(torch.zeros(1, self.hidden_size))

n_hidden = 128
rnn = RNN(n_letters, n_hidden, n_categories)
```   

The RNN formula is:   
![image](https://user-images.githubusercontent.com/3640001/34774225-800e51fa-f5dc-11e7-9180-996be8e52a8f.png)

Which is implemented by these lines:    
```python    
        combined = torch.cat((input, hidden), 1)
        hidden = self.i2h(combined)
```    

However, these lines:    
```python    
        output = self.i2o(combined)   
        output = self.softmax(output)
```   
Are trying to project into the classification space. However, the self.i2o operates on the combined output instead of the ht output.    

This implementation uses the wrong formula:    
![image](https://user-images.githubusercontent.com/3640001/34774199-6772b9b0-f5dc-11e7-8cc0-c1e5cdb8a650.png)

But the correct formula is:      
![image](https://user-images.githubusercontent.com/3640001/34774212-7358d110-f5dc-11e7-8e3a-be07e1e4c3fc.png)

     
Which can be implemented as:   
```python    
    def forward(self, input, hidden):
        combined = torch.cat((input, hidden), 1)
        hidden = self.i2h(combined)
        output = self.i2o(hidden) # this line changed   (bc hidden = ht. combined = ht-1)
        output = self.softmax(output)
        return output, hidden
```    

Basically, the current implementation does this:    
![image](https://user-images.githubusercontent.com/3640001/34774753-93dbf80c-f5de-11e7-8429-70a510c3cc11.png)

But it really should do this:    
![image](https://user-images.githubusercontent.com/3640001/34774734-7abeb0e4-f5de-11e7-885f-360b7cfaec93.png)

**Comment by ayush1999** [comment]
@williamFalcon Small question. You mentioned that 
```
combined = torch.cat((input, hidden), 0)
hidden = self.i2h(combined)
```

Performs the RNN formula. I didn't get how concatenating the input and hidden tensors is useful in this scenario. Why the need for concatenation?

**Comment by EmreOzkose** [comment]
I think it should be changed. The implemented formula is wrong?

**Comment by Lguyogiro** [comment]
> @williamFalcon Small question. You mentioned that
> 
> ```
> combined = torch.cat((input, hidden), 0)
> hidden = self.i2h(combined)
> ```
> 
> Performs the RNN formula. I didn't get how concatenating the input and hidden tensors is useful in this scenario. Why the need for concatenation?

I know this is an old issue/comment....but you don't "need" to concatenate the input and hidden tensors, this is just a way to calculate the same thing without having two separate weights tensors (the standard formulae typically assume two weights matrices, see below) 
![image](https://user-images.githubusercontent.com/7957321/194729841-fbbe6545-4334-43c3-b21e-7c10fd8f701c.png)


Also I just wanted to chime in. I agree that this should be changed.
